Reviewed the Creating Factories section for language a… (#620)
codenvy/che-docs: src/main/pages/portable-workspaces/creating-factories.adoc

Creating a factory in the IDE in a running workspace generates a factory to replicate that workspace including the runtime and project settings.

A factory created from the dashboard is persisted on {{ site.product_mini_name }} and retained when upgrading to a newer version.

To create a factory in the IDE:

. In the IDE, click *Workspace* > *Create Factory*.

Sample factory: https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j[https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j]

[id="create-a-factory-based-on-a-repo"]
== Creating a factory based on a repository

URL factories work with GitHub and GitLab repositories. By using URL factories, the project referenced by the URL is automatically imported.

To create a factory based on a repository:

. Specify the repository URL. Ensure that you store the configuration in the repository.

Sample factories:

http://che.openshift.io/f?url=https://github.com/eclipse/che +
http://che.openshift.io/f?url=https://github.com/eclipse/che/tree/language-server +
http://che.openshift.io/f?url=https://gitlab.com/benoitf/simple-project

The factory URL can include a branch or a subdirectory. Following are examples of optional parameters:

> * `?url=https://github.com/eclipse/che` Che is imported with the `master` branch.

Reviewer: Perhaps a definition list would be more appropriate here? We're essentially showing examples of optional parameters with an explanation of what they demonstrate.

* `?url=https://github.com/eclipse/che/tree/5.0.0` Che is imported by using the `5.0.0` branch.

> * `?url=https://github.com/eclipse/che/tree/5.0.0/dashboard` subdirectory dashboard is imported by using the `5.0.0` branch.

Reviewer: `dashboard` should be styled as a directory name.

folder: portable-workspaces
---

[id="create-a-factory-in-the-dashboard"]

> == Creating a factory on the dashboard

Reviewer 2: ```suggestion
== Creating a factory in the dashboard
```

> You can create a factory based on an existing workspace. You can also create factories based on a template or by pasting in a `factory.json` file and then generating a factory URL using the Che CLI or API. To learn more about the JSON structure and options, see link:factories_json_reference.html[Factory JSON reference].

Reviewer 2: ```suggestion
You can create a factory based on an existing workspace. You can also create factories based on a template or by pasting in a `.factory.json` file and then generating a factory URL using the Che CLI or API. To learn more about the JSON structure and options, see link:factories_json_reference.html[Factory JSON reference].
```

A factory created from the dashboard persists in the Che database and is retained when upgrading to a newer version.

To create a factory on the dashboard:

. In the IDE, click *Dashboard* > *Factories* > *Create Factory*.

Sample factory: https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j[https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j]

[id="create-a-factory-in-the-ide"]
== Creating a factory in the IDE

Creating a factory in the IDE in a running workspace generates a factory to replicate that workspace including the runtime and project settings.

> A factory created from the dashboard is persisted on {{ site.product_mini_name }} and retained when upgrading to a newer version.

Reviewer 2: ```suggestion
A factory created from the dashboard is persisted on Che and retained when upgrading to a newer version.
```

To create a factory in the IDE:

. In the IDE, click *Workspace* > *Create Factory*.

> Sample factory: https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j[https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j]

Reviewer 2: ```suggestion
Sample factory: https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j[https://che.openshift.io/f?id=factorymtyoro1y0qt8tq2j].
```

[id="create-a-factory-based-on-a-repo"]
== Creating a factory based on a repository

URL factories work with GitHub and GitLab repositories. By using URL factories, the project referenced by the URL is automatically imported.

To create a factory based on a repository:

. Specify the repository URL. Ensure that you store the configuration in the repository.

Sample factories:

> http://che.openshift.io/f?url=https://github.com/eclipse/che +

Reviewer 2: ```suggestion
* http://che.openshift.io/f?url=https://github.com/eclipse/che
```

> http://che.openshift.io/f?url=https://github.com/eclipse/che/tree/language-server +

Reviewer 2: ```suggestion
* http://che.openshift.io/f?url=https://github.com/eclipse/che/tree/language-server
```

> http://che.openshift.io/f?url=https://gitlab.com/benoitf/simple-project

Reviewer 2: ```suggestion
* http://che.openshift.io/f?url=https://gitlab.com/benoitf/simple-project
```

The factory URL can include a branch or a subdirectory. Following are examples of optional parameters:

`?url=https://github.com/eclipse/che`::
	Che is imported with the `master` branch.

`?url=https://github.com/eclipse/che/tree/5.0.0`:: 
	Che is imported by using the `5.0.0` branch.

`?url=https://github.com/eclipse/che/tree/5.0.0/dashboard`:: 

> 	Subdirectory `dashboard` is imported by using the `5.0.0` branch.

Reviewer 2: ```suggestion
	Subdirectory `dashboard/` is imported by using the `5.0.0` branch.
```

[id="customizing-url-factories"]
=== Customizing URL factories

> Following are two ways to customize the runtime and configuration:

Reviewer 2: ```suggestion
The following are two ways to customize the runtime and configuration:
```

* Customizing only the runtime

> * Dockerfile works on Docker infrastructure only

Reviewer 2: If this is the 2nd way to customize the runtime, then where is it described below? I only see "Customizing the project and runtime".

> Providing a `.factory.dockerfile` inside the repository signals to the {{ site.product_mini_name }} URL factory to use this Dockerfile for the workspace agent runtime. By default, imported projects are set to a `blank` project type. You can also set the project type in the `.factory.json` or workspace definition that the factory inherits from.

Reviewer 2: ```suggestion
Providing a `.factory.dockerfile` inside the repository signals to the URL factory to use this Dockerfile for the workspace agent runtime. By default, imported projects are set to a `blank` project type. You can also set the project type in the `.factory.json` file or in the workspace definition that the factory inherits from.
```

> *Customizing the project and runtime*

Reviewer 2: ```suggestion
==== Customizing the project and runtime
```

> Providing a `.factory.json` file inside the repository will signal to Che URL factory to configure the project and runtime according to this configuration file. When a `.factory.json` file is stored inside the repository, any `Dockerfile` content is ignored because the workspace runtime configuration is defined inside the JSON file.

Reviewer 2: ```suggestion
Providing a `.factory.json` file inside the repository signals to the Che URL factory to configure the project and runtime according to this configuration file. When a `.factory.json` file is stored inside the repository, any `Dockerfile` content is ignored because the workspace runtime configuration is defined inside the JSON file.
```

[id="factory-policies"]

> == Factory policies

Reviewer 2: ```suggestion
== Configuring factory policies
```

== Limitations

*Referer*::
  Checks the hostname of the acceptor and only allows the factory to execute if there is a match.
*Since* & *Until*::
  Defines the time window in which the factory can be activated. For example, instructors who want to create an exercise that can only be accessed for two hours should set these properties.

[id="multiplicity"]
== Multiplicity

Multiplicity defines the number of workspaces that can be created from the factory.

*Multiple Workspaces: perClick*::
  Every click of the factory URL generates a different workspace, each with its own identifier, name, and resources.
*Single Workspace: perUser*::
  Exactly one workspace is generated for each unique user that clicks on the factory URL. If the workspace has previously been generated, the existing workspace is reopened.

To configure factory policies, see link:factories_json_reference.html[JSON reference].

[id="ide-customization"]
== IDE customization

You can instruct the factory to invoke a series of IDE actions based on events in the lifecycle of the workspace.

[id="lifecycle-events"]
== Lifecycle events

The lifecycle of the workspace is defined by the following events:

> * `onAppLoaded`: Triggered when the IDE is loaded.

Reviewer: Again, a definition list may be more appropriate here.

Each event type has a set of actions that can be triggered. There is no ordering of actions executed when you provide a list; {{ site.product_mini_name }} will asynchronously invoke multiple actions if appropriate.

[id="factory-actions"]
== Factory actions

Following is the list of all possible actions that can be configured with your factory.

*Run Command*::
  _Description:_ Specify the name of the command to invoke after the IDE is loaded. +
  _Associated Event:_ `onProjectsLoaded`
*Open File*::
  _Description:_ Open project files in the editor. Optionally, define the line to be highlighted. +
  _Associated Event:_ `onProjectsLoaded`
*Open a Welcome Page*::
  _Description:_ Customize content of a welcome panel displayed when the workspace is loaded. +
  _Associated Event:_ `onAppLoaded`
*Warm on Uncommitted Changes*::
  _Description:_ Opens a warning pop-up window when the user closes the browser tab with a project that has uncommitted changes. +
  _Associated Event:_ `onAppClosed`

To configure factory actions, see link:factories_json_reference.html#ide-customization[Factory JSON reference].

[id="find-and-replace"]
== Finding and replacing variables

> To expose sensitive information like passwords, URLs, account names, API keys, in the source code, factories allow you to replace variables or placeholders with real values. To find and replace a value, you can use the `run` command during an `onProjectsLoaded` event. You can use `sed`, `awk` or other tools available in your workspace environment.

Reviewer: The opening sentence is awkward and makes it look like "in the source code" is part of the list "passwords, URLs, account names, API keys".

Author: Changed to "Factories allow you to replace variables or placeholders with real values to expose sensitive information (passwords, URLs, account names, API keys) in the source code ". Better?

> == Limitations

Reviewer 2: ```suggestion
=== Setting factory limitations
```

*Referer*::
  Checks the hostname of the acceptor and only allows the factory to execute if there is a match.
*Since* & *Until*::
  Defines the time window in which the factory can be activated. For example, instructors who want to create an exercise that can only be accessed for two hours should set these properties.

[id="multiplicity"]

> == Multiplicity

Reviewer 2: ```suggestion
=== Setting factory multiplicity
```

Multiplicity defines the number of workspaces that can be created from the factory.

*Multiple Workspaces: perClick*::
  Every click of the factory URL generates a different workspace, each with its own identifier, name, and resources.
*Single Workspace: perUser*::
  Exactly one workspace is generated for each unique user that clicks on the factory URL. If the workspace has previously been generated, the existing workspace is reopened.

> To configure factory policies, see link:factories_json_reference.html[JSON reference].

Reviewer 2: The orig. was better. The edit also changed meaning.

[id="ide-customization"]

> == IDE customization

Reviewer 2: ```suggestion
== Customizing the IDE
```

You can instruct the factory to invoke a series of IDE actions based on events in the lifecycle of the workspace.

[id="lifecycle-events"]
== Lifecycle events

The lifecycle of the workspace is defined by the following events:

`onAppLoaded`::
	Triggered when the IDE is loaded.

`onProjectsLoaded`::
	Triggered when the workspace and all projects have been activated.

`onAppClosed`::
	Triggered when the IDE is closed.

> Each event type has a set of actions that can be triggered. There is no ordering of actions executed when you provide a list; {{ site.product_mini_name }} will asynchronously invoke multiple actions if appropriate.

Reviewer 2: ```suggestion
Each event type has a set of actions that can be triggered. There is no ordering of actions executed when you provide a list; Che asynchronously invokes multiple actions if appropriate.
```

[id="factory-actions"]
== Factory actions

> Following is the list of all possible actions that can be configured with your factory.

Reviewer 2: ```suggestion
The following is a list of all possible actions that can be configured with your factory.
```

*Run Command*::
  _Description:_ Specify the name of the command to invoke after the IDE is loaded. +
  _Associated Event:_ `onProjectsLoaded`
*Open File*::

>   _Description:_ Open project files in the editor. Optionally, define the line to be highlighted. +

Reviewer 2: ```suggestion
  _Description_::: Open project files in the editor. Optionally, define the line to be highlighted.
```

>   _Associated Event:_ `onProjectsLoaded`

Reviewer 2: ```suggestion
  _Associated Event_::: `onProjectsLoaded`
```

*Open a Welcome Page*::

>   _Description:_ Customize content of a welcome panel displayed when the workspace is loaded. +

Reviewer 2: ```suggestion
  _Description_::: Customize content of a welcome panel displayed when the workspace is loaded.
```

>   _Associated Event:_ `onAppLoaded`

Reviewer 2: ```suggestion
  _Associated Event_::: `onAppLoaded`
```

*Warm on Uncommitted Changes*::

>   _Description:_ Opens a warning pop-up window when the user closes the browser tab with a project that has uncommitted changes. +

Reviewer 2: ```suggestion
  _Description_::: Opens a warning pop-up window when the user closes the browser tab with a project that has uncommitted changes.
```

>   _Associated Event:_ `onAppClosed`

Reviewer 2: ```suggestion
  _Associated Event_::: `onAppClosed`
```

> To configure factory actions, see link:factories_json_reference.html#ide-customization[Factory JSON reference].

Reviewer 2: Orig. better.

[id="find-and-replace"]
== Finding and replacing variables

> Factories allow you to replace variables or placeholders with real values to expose sensitive information (passwords, URLs, account names, API keys) in the source code. To find and replace a value, you can use the `run` command during an `onProjectsLoaded` event. You can use `sed`, `awk` or other tools available in your workspace environment.

Reviewer 2: You're changing the meaning. Orig. talks about **not** exposing secrets.
```suggestion
Factories make it possible to replace variables or placeholders in the source code -- used to avoid exposing sensitive information (passwords, URLs, account names, API keys) -- with real values. To find and replace a value, you can use the `run` command during an `onProjectsLoaded` event. You can use `sed`, `awk`, or other tools available in your workspace environment.
```

> For a sample of how to configure finding and replacing a value, see the link:factories_json_reference.html#action-find-and-replace[Factory JSON reference] section. Alternatively, you can add IDE actions in *Factory* tab, on the user *Dashboard*.

Reviewer 2: ```suggestion
For a sample of how to configure finding and replacing a value, see the link:factories_json_reference.html#action-find-and-replace[Factory JSON reference] section. Alternatively, you can add IDE actions in the *Factory* tab, on the user *Dashboard*.
```

Use https://www.gnu.org/software/sed/manual/html_node/Regular-Expressions.html[regular expressions] in `sed`, both in find-replace and file-file type patterns.

[id="pull-request-workflow"]
== Pull request workflow

> Factories can be configured with a dedicated pull request workflow. The PR workflow handles local and remote branching, forking, and issuing the pull request. Pull requests generated from within {{ site.product_mini_name }} have another factory placed into the comments of the pull requests that a PR reviewer can use to quickly start the workspace.

Reviewer 2: ```suggestion
Factories can be configured with a dedicated pull request workflow. The PR workflow handles local and remote branching, forking, and issuing the pull request. Pull requests generated from within Che have another factory placed into the comments of the pull requests that a PR reviewer can use to quickly start the workspace.
```

[id="nest-steps"]

> == Nest Steps

Reviewer 2: ```suggestion
== Nest steps
```

Reviewer: This looks like a typo to me -- was "Next steps" intended?
